File: 8. String to Integer (atoi)/cpp/solution.cpp
class Solution {
public:
    int myAtoi(string s) {
        double solution = 0;
        int i = 0;
        bool isNegative = false;
        while(s[i] == ' '){
            i++;
        }
        if(s[i] == '-' || s[i] == '+'){
            if(s[i] == '-') isNegative = true;
            i++;
        }
        while(s[i] - '0' <= 9 && s[i] - '0' >= 0){
            solution*=10;
            solution += s[i] - '0';
            i++;
        }
        solution = solution * (isNegative ? -1 : 1);
        if(solution > INT_MAX) return INT_MAX;
        else if(solution < INT_MIN) return INT_MIN;
        return solution;
    }
};

File: 8. String to Integer (atoi)/python/solution.py
class Solution(object):
    def myAtoi(self, s):
        s = s.lstrip()

        match = re.match(r'^([+-]?\d+)', s)
        if not match:
            return 0

        num = int(match.group(1))

        INT_MAX = 2**31 - 1
        INT_MIN = -2**31

        if num > INT_MAX:
            return INT_MAX
        if num < INT_MIN:
            return INT_MIN
        return num
        